File: mobile/android/app/src/main/java/com/mesh/wallet/data/tron/TronApiService.kt
package com.mesh.wallet.data.tron

import com.mesh.wallet.core.config.TronConfiguration
import com.mesh.wallet.domain.model.TransactionDirection
import com.mesh.wallet.domain.model.WalletTransaction
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.math.BigDecimal
import java.math.RoundingMode

data class TronAccountResources(
    val energyRemaining: Long = 0,
    val bandwidthRemaining: Long = 0,
    val hasEnoughTrxForFees: Boolean = false
)

object TronApiService {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun fetchUsdtBalance(address: String): BigDecimal? {
        return fetchUsdtBalanceFromTrc20(address) ?: fetchUsdtBalanceFromAccount(address)
    }

    suspend fun fetchAccountBalance(address: String): com.mesh.wallet.domain.model.TronAccountBalance {
        val trx = fetchTrxBalance(address)
        val usdt = fetchUsdtBalance(address) ?: BigDecimal.ZERO
        val txCount = fetchTransactionCount(address)
        return com.mesh.wallet.domain.model.TronAccountBalance(
            trxBalance = trx,
            usdtBalance = usdt,
            transactionCount = txCount
        )
    }

    suspend fun isAccountActivated(address: String): Boolean {
        return runCatching {
            val raw = TronApiClient.post("/wallet/getaccount", TronApiClient.accountBody(address))
            val response = json.decodeFromString<AccountResponse>(raw)
            response.createTime != null
        }.getOrDefault(true)
    }

    suspend fun fetchAccountResources(address: String): TronAccountResources {
        val raw = TronApiClient.post(
            "/wallet/getaccountresource",
            TronApiClient.accountBody(address)
        )
        val decoded = json.decodeFromString<AccountResourceResponse>(raw)
        val energy = (decoded.energyLimit ?: 0L) - (decoded.energyUsed ?: 0L)
        val freeNet = (decoded.freeNetLimit ?: 0L) - (decoded.freeNetUsed ?: 0L)
        val net = (decoded.netLimit ?: 0L) - (decoded.netUsed ?: 0L)
        val bandwidth = maxOf(freeNet, net)
        val trx = fetchTrxBalance(address)
        return TronAccountResources(
            energyRemaining = maxOf(energy, 0),
            bandwidthRemaining = maxOf(bandwidth, 0),
            hasEnoughTrxForFees = trx >= 1.0 || energy > 0
        )
    }

    suspend fun fetchUsdtTransactions(address: String, limit: Int = 20): List<WalletTransaction> {
        val raw = TronApiClient.get(
            path = "/v1/accounts/$address/transactions/trc20",
            query = mapOf(
                "limit" to limit.toString(),
                "contract_address" to TronConfiguration.usdtContractAddress,
                "only_confirmed" to "true"
            )
        )
        val decoded = json.decodeFromString<Trc20HistoryResponse>(raw)
        return decoded.data.mapNotNull { item -> mapHistoryItem(item, address) }
    }

    private suspend fun fetchTrxBalance(address: String): Double {
        val raw = TronApiClient.post("/wallet/getaccount", TronApiClient.accountBody(address))
        val response = json.decodeFromString<AccountResponse>(raw)
        val sun = response.balance ?: 0L
        return sun / 1_000_000.0
    }

    private suspend fun fetchTransactionCount(address: String): Int {
        return runCatching {
            val raw = TronApiClient.get("/v1/accounts/$address")
            json.decodeFromString<AccountListResponse>(raw).data.firstOrNull()?.transactions ?: 0
        }.getOrDefault(0)
    }

    private suspend fun fetchUsdtBalanceFromTrc20(address: String): BigDecimal? {
        return runCatching {
            val raw = TronApiClient.get(
                path = "/v1/accounts/$address/trc20/balance",
                query = mapOf(
                    "contract_address" to TronConfiguration.usdtContractAddress,
                    "limit" to "1"
                )
            )
            val decoded = json.decodeFromString<Trc20BalanceResponse>(raw)
            val tokenValue = decoded.data.firstOrNull()?.get(TronConfiguration.usdtContractAddress)
            parseUsdt(tokenValue ?: "0")
        }.getOrNull()
    }

    private suspend fun fetchUsdtBalanceFromAccount(address: String): BigDecimal? {
        return runCatching {
            val raw = TronApiClient.get("/v1/accounts/$address")
            val decoded = json.decodeFromString<AccountListResponse>(raw)
            for (token in decoded.data.firstOrNull()?.trc20.orEmpty()) {
                val value = token[TronConfiguration.usdtContractAddress]
                if (value != null) return parseUsdt(value)
            }
            BigDecimal.ZERO
        }.getOrNull()
    }

    private fun parseUsdt(raw: String): BigDecimal {
        val value = raw.toBigDecimalOrNull() ?: BigDecimal.ZERO
        return value.divide(BigDecimal.TEN.pow(TronConfiguration.TOKEN_DECIMALS), 6, RoundingMode.HALF_UP)
    }

    private fun mapHistoryItem(item: Trc20HistoryItem, walletAddress: String): WalletTransaction? {
        val amount = parseUsdt(item.value ?: return null)
        if (amount <= BigDecimal.ZERO) return null
        val from = item.from ?: return null
        val to = item.to ?: return null
        val direction = if (from.equals(walletAddress, ignoreCase = true)) {
            TransactionDirection.OUTGOING
        } else {
            TransactionDirection.INCOMING
        }
        val counterparty = if (direction == TransactionDirection.OUTGOING) to else from
        return WalletTransaction(
            id = item.transactionId ?: item.eventIndex?.toString() ?: return null,
            txId = item.transactionId,
            direction = direction,
            amount = amount,
            counterpartyAddress = counterparty,
            timestamp = item.blockTimestamp ?: System.currentTimeMillis()
        )
    }
}

@Serializable
private data class AccountResponse(
    val balance: Long? = null,
    @SerialName("create_time") val createTime: Long? = null
)

@Serializable
private data class AccountResourceResponse(
    @SerialName("EnergyLimit") val energyLimit: Long? = null,
    @SerialName("EnergyUsed") val energyUsed: Long? = null,
    @SerialName("freeNetLimit") val freeNetLimit: Long? = null,
    @SerialName("freeNetUsed") val freeNetUsed: Long? = null,
    @SerialName("NetLimit") val netLimit: Long? = null,
    @SerialName("NetUsed") val netUsed: Long? = null
)

@Serializable
private data class AccountListResponse(val data: List<AccountData> = emptyList())

@Serializable
private data class AccountData(
    val transactions: Int? = null,
    val trc20: List<Map<String, String>>? = null
)

@Serializable
private data class Trc20BalanceResponse(val data: List<Map<String, String>> = emptyList())

@Serializable
private data class Trc20HistoryResponse(val data: List<Trc20HistoryItem> = emptyList())

@Serializable
private data class Trc20HistoryItem(
    @SerialName("transaction_id") val transactionId: String? = null,
    val from: String? = null,
    val to: String? = null,
    val value: String? = null,
    @SerialName("block_timestamp") val blockTimestamp: Long? = null,
    @SerialName("event_index") val eventIndex: Int? = null
)

File: mobile/ios/Mesh/Services/Tron/TronAPIService.swift
import Foundation

enum TronAPIService {
    static func fetchAccountBalance(address: String) async -> TronAccountBalance {
        async let trxSignals = fetchTRXBalance(address: address)
        async let usdtBalance = fetchUSDTBalance(address: address)
        async let txCount = fetchTransactionCount(address: address)
        let (trx, _) = await trxSignals
        return TronAccountBalance(
            trxBalance: trx,
            usdtBalance: await usdtBalance ?? 0,
            transactionCount: await txCount
        )
    }

    /// Returns `nil` when TronGrid is unreachable or rate-limited (do not treat as zero balance).
    static func fetchUSDTBalance(address: String) async -> Double? {
        if let balance = await fetchUSDTBalanceFromTRC20Endpoint(address: address) {
            return balance
        }
        return await fetchUSDTBalanceFromAccountEndpoint(address: address)
    }

    /// `false` when the address has never received TRX (cannot receive TRC-20 yet).
    static func isAccountActivated(address: String) async -> Bool {
        let normalized = TronAddressCodec.normalizedBase58(address)
            ?? address.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty else { return false }

        if let viaV1 = await isAccountActivatedViaV1(normalized) {
            return viaV1
        }
        return await isAccountActivatedViaGetAccount(normalized)
    }

    /// TronGrid lists only on-chain accounts; empty `data` means the address was never funded.
    private static func isAccountActivatedViaV1(_ address: String) async -> Bool? {
        do {
            let data = try await TronAPIClient.get(path: "/v1/accounts/\(address)")
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let rows = json["data"] as? [[String: Any]]
            else {
                return nil
            }
            guard let account = rows.first else {
                return false
            }
            return accountJSONIndicatesActivated(account)
        } catch {
            return nil
        }
    }

    private static func isAccountActivatedViaGetAccount(_ address: String) async -> Bool {
        do {
            let data = try await TronAPIClient.post(
                path: "/wallet/getaccount",
                jsonBody: ["address": address, "visible": true]
            )
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return false
            }
            return accountJSONIndicatesActivated(json)
        } catch {
            return false
        }
    }

    private static func accountJSONIndicatesActivated(_ json: [String: Any]) -> Bool {
        if json.isEmpty {
            return false
        }
        if positiveInt64(json["create_time"]) != nil {
            return true
        }
        let balanceSun = int64Value(json["balance"]) ?? 0
        return balanceSun >= 1_000_000
    }

    private static func positiveInt64(_ value: Any?) -> Int64? {
        guard let parsed = int64Value(value), parsed > 0 else { return nil }
        return parsed
    }

    private static func int64Value(_ value: Any?) -> Int64? {
        switch value {
        case let number as Int:
            return Int64(number)
        case let number as Int64:
            return number
        case let number as Double:
            return Int64(number)
        case let number as NSNumber:
            return number.int64Value
        case let text as String:
            let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty else { return nil }
            return Int64(trimmed)
        default:
            return nil
        }
    }

    static func fetchAccountResources(address: String) async throws -> TronAccountResources {
        let data = try await TronAPIClient.post(
            path: "/wallet/getaccountresource",
            jsonBody: ["address": address, "visible": true]
        )
        let decoded = try JSONDecoder().decode(TronAccountResourceResponse.self, from: data)
        let energy = (decoded.EnergyLimit ?? 0) - (decoded.EnergyUsed ?? 0)
        let freeNet = (decoded.freeNetLimit ?? 0) - (decoded.freeNetUsed ?? 0)
        let net = (decoded.NetLimit ?? 0) - (decoded.NetUsed ?? 0)
        let bandwidth = max(freeNet, net)
        let balance = await fetchTRXBalance(address: address).0
        return TronAccountResources(
            energyRemaining: Int64(max(energy, 0)),
            bandwidthRemaining: Int64(max(bandwidth, 0)),
            hasEnoughTRXForFees: balance >= 1.0 || energy > 0
        )
    }

    static func fetchUSDTTransactions(
        address: String,
        limit: Int = 20
    ) async throws -> [TronUSDTTransaction] {
        let query = [
            URLQueryItem(name: "limit", value: String(limit)),
            URLQueryItem(name: "contract_address", value: TronConfiguration.usdtContractAddress),
            URLQueryItem(name: "only_confirmed", value: "true")
        ]
        let data = try await TronAPIClient.get(
            path: "/v1/accounts/\(address)/transactions/trc20",
            query: query
        )
        let decoded = try JSONDecoder().decode(TronTRC20HistoryResponse.self, from: data)
        return decoded.data.compactMap { item in
            mapHistoryItem(item, walletAddress: address)
        }
    }

    private static func fetchTRXBalance(address: String) async -> (Double, Int) {
        do {
            let data = try await TronAPIClient.post(
                path: "/wallet/getaccount",
                jsonBody: ["address": address, "visible": true]
            )
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return (0, 0)
            }
            let raw = (json["balance"] as? Double) ?? Double(json["balance"] as? Int ?? 0)
            return (raw / 1_000_000.0, 0)
        } catch {
            return (0, 0)
        }
    }

    private static func fetchTransactionCount(address: String) async -> Int {
        do {
            let data = try await TronAPIClient.get(path: "/v1/accounts/\(address)")
            let decoded = try JSONDecoder().decode(TronGridAccountResponse.self, from: data)
            return decoded.data.first?.transactions ?? 0
        } catch {
            return 0
        }
    }

    /// Preferred: TronGrid TRC-20 balance endpoint (reliable for USDT).
    private static func fetchUSDTBalanceFromTRC20Endpoint(address: String) async -> Double? {
        do {
            let query = [
                URLQueryItem(name: "contract_address", value: TronConfiguration.usdtContractAddress),
                URLQueryItem(name: "limit", value: "1"),
            ]
            let data = try await TronAPIClient.get(
                path: "/v1/accounts/\(address)/trc20/balance",
                query: query
            )
            let decoded = try JSONDecoder().decode(TronTRC20BalanceResponse.self, from: data)
            for tokenMap in decoded.data {
                if let raw = tokenMap[TronConfiguration.usdtContractAddress] {
                    return parseUSDTBalanceString(raw)
                }
            }
            return 0
        } catch {
            return nil
        }
    }

    /// Fallback for older account payloads that embed `trc20` on `/v1/accounts/{address}`.
    private static func fetchUSDTBalanceFromAccountEndpoint(address: String) async -> Double? {
        do {
            let data = try await TronAPIClient.get(path: "/v1/accounts/\(address)")
            let decoded = try JSONDecoder().decode(TronGridAccountResponse.self, from: data)
            guard let trc20Tokens = decoded.data.first?.trc20 else { return 0 }

            for tokenMap in trc20Tokens {
                if let rawBalance = tokenMap[TronConfiguration.usdtContractAddress] {
                    return parseUSDTBalanceString(rawBalance)
                }
            }
            return 0
        } catch {
            return nil
        }
    }

    private static func parseUSDTBalanceString(_ raw: String) -> Double {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return 0 }
        if let units = UInt64(trimmed) {
            let decimal = TronAmountEncoder.smallestUnitsToUSDT(units)
            return (decimal as NSDecimalNumber).doubleValue
        }
        return Double(trimmed).map { $0 / pow(10, Double(TronConfiguration.tokenDecimals)) } ?? 0
    }

    private static func mapHistoryItem(
        _ item: TronTRC20HistoryItem,
        walletAddress: String
    ) -> TronUSDTTransaction? {
        guard let txID = item.transaction_id,
              let from = item.from,
              let to = item.to,
              let valueRaw = item.value,
              let timestampMs = item.block_timestamp
        else { return nil }

        let smallestUnits = UInt64(valueRaw) ?? UInt64(Double(valueRaw) ?? 0)
        let amount = TronAmountEncoder.smallestUnitsToUSDT(smallestUnits)
        // Classify relative to the address we queried — required for sends between own receive slots.
        let direction: TronTransactionDirection
        if TronAddressCodec.matches(from, walletAddress) {
            direction = .outgoing
        } else if TronAddressCodec.matches(to, walletAddress) {
            direction = .incoming
        } else {
            return nil
        }
        return TronUSDTTransaction(
            id: txID,
            txID: txID,
            fromAddress: from,
            toAddress: to,
            amount: amount,
            timestamp: Date(timeIntervalSince1970: TimeInterval(timestampMs) / 1000),
            direction: direction
        )
    }

    private static func addressesMatch(_ lhs: String, _ rhs: String) -> Bool {
        TronAddressCodec.matches(lhs, rhs)
    }
}

private struct TronGridAccountResponse: Decodable {
    let data: [TronGridAccount]
}

private struct TronGridAccount: Decodable {
    let trc20: [[String: String]]?
    let transactions: Int?
}

private struct TronAccountResourceResponse: Decodable {
    let EnergyLimit: Int64?
    let EnergyUsed: Int64?
    let NetLimit: Int64?
    let NetUsed: Int64?
    let freeNetLimit: Int64?
    let freeNetUsed: Int64?
}

private struct TronTRC20BalanceResponse: Decodable {
    let data: [[String: String]]
}

private struct TronTRC20HistoryResponse: Decodable {
    let data: [TronTRC20HistoryItem]
}

private struct TronTRC20HistoryItem: Decodable {
    let transaction_id: String?
    let from: String?
    let to: String?
    let value: String?
    let block_timestamp: Int64?
}
